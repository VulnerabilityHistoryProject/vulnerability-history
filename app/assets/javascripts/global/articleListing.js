// import $ from 'jquery'
// import _ from 'lodash'
import {vhpMarkdown} from './vhpMarkdown'

export default function articleListing(jsonData, container = '#article-listing') {
  const blurbCounts = {
    '#blurb-1big-template' : 1,
    '#blurb-4even-template' : 4,
    '#blurb-1wide-template' : 1,
    '#blurb-1big1small-template' : 2,
  }
  const numTemplates = _.size(blurbCounts);
  const bMax = _.reduce(blurbCounts, (max, v, _k) => max = v > max ? v : max, 0)
  let articlesLeft = jsonData.length;
  let i = 0;
  while(articlesLeft > bMax){
    const templateID = Object.keys(blurbCounts)[i];
    addBlurbRow(container, templateID)
    i = (i + 1) % numTemplates;
    articlesLeft -= blurbCounts[templateID];
  }
  //for the last few we add a single wide row.
  _.times(articlesLeft, () => addBlurbRow(container, '#blurb-1wide-template'))

  $('.hidden').remove() //Remove leftover blurb templates

  let blurbs = $('.blurb-text')
  let blurbArts = $('.blurb-art')
  for (let i = 0; i < jsonData.length; i++) {
    let article = jsonData[i]
    $(blurbArts[i]).html(`<div class="vhp-art-${article.art}"/>`)
    blurbs[i].innerHTML = vhpMarkdown(`
### ${article.title} ###
${article.blurb}
`) + `<a href="/articles/${article.slug}" class="button read-more">Read more</a>`
  }
}

export function articleListingOneSize(jsonData, container = '#article-listing-same') {
  _.times(jsonData.length,() => addBlurbRow(container, '#blurb-1wide-template'))
  let blurbs = $(container + ' .blurb-text');
  let blurbArts = $(container + ' .blurb-art');
  for (let i = 0; i < jsonData.length; i++) {
    let article = jsonData[i];
    $(blurbArts[i]).html(`<div class="vhp-art-${article.art}"/>`);
    blurbs[i].innerHTML = vhpMarkdown(`
### ${article.title} ###
${article.blurb}
[Full Article](/articles/${article.slug})
    `);
  }
}

function addBlurbRow(container, templateID) {
  const blurbRow = $(templateID).clone();
  blurbRow.removeAttr('id').removeClass('hidden');
  $(container).append(blurbRow).append(`<hr/>`);
}
