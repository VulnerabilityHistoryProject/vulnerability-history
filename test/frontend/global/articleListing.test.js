import articleListing from "../../../app/frontend/global/articleListing"
import articleListingOneSize from "../../../app/frontend/global/articleListing"
import $ from 'jquery'


const articles = [
    {
        art: "snowflake",
        author: "Andy Meneely",
        blurb: "Don't just think about code complexity, think about *input* complexity.↵↵Computers only understand numbers,",
        cves: [null],
        filepaths: [],
        fulltext: "↵The full text of this article can be found at the [:tag: Complex Inputs](/tags/complex-inputs) tag page.↵",
        id: 1,
        slug: "beware-complex-inputs",
        title: "Beware of complex inputs"
    },
    {
        art: "question",
        author: "Andy Meneely",
        blurb: "Why do we talk about things that have not been \"proven\" in research?↵↵Because science is a work in progress.",
        cves: [null],
        filepaths: [],
        fulltext: "↵↵_You can't talk about that, it's not proven!_↵↵If we waited for the evidence to be crystal clear,",
        id: 2,
        slug: "open-questions",
        title: "We Love Open Questions"
    },
    {
        art: "refactor",
        author: "Andy Meneely",
        blurb: "You can't secure it if you can't maintain it.↵↵**Refactoring** is the practice of improving the maintainability",
        cves: [null],
        filepaths: [],
        fulltext: "↵_You can't secure it if you can't maintain it._↵↵A widespread practice, refactoring is useful for improving",
        id: 3,
        slug: "refactoring",
        title: "Always Be Refactoring"
    }
]

// TODO: Issue 826 No Return
// SNAPSHOT
test("Article Listing", () => {
    document.body.innerHTML =
        '<div id="article-listing">' +
            '<div class="blurb-text" cell medium-8"></div>' +
            '<div class="blurb-art" cell medium-3"></div>' +
            '<div class="blurb-text" cell medium-8"></div>' +
            '<div class="blurb-art" cell medium-3"></div>' +
            '<div class="blurb-text" cell medium-8"></div>' +
            '<div class="blurb-art" cell medium-3"></div>' +
        '</div>';
    articleListing(articles, '#article-listing');
    expect($('#article-listing').html()).toMatchSnapshot();
});

test("Display All Articles as One Size", () => {
    document.body.innerHTML =
        '<div id="article-listing-same">' +
            '<div class="blurb-text" cell medium-8"></div>' +
            '<div class="blurb-art" cell medium-3"></div>' +
            '<div class="blurb-text" cell medium-8"></div>' +
            '<div class="blurb-art" cell medium-3"></div>' +
            '<div class="blurb-text" cell medium-8"></div>' +
            '<div class="blurb-art" cell medium-3"></div>' +
        '</div>';
    articleListingOneSize(articles, '#article-listing-same');
    expect($('#article-listing-same').html()).toMatchSnapshot();
});
