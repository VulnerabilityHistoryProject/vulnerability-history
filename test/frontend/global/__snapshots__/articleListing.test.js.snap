// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Article Listing 1`] = `
"<div class=\\"blurb-text\\" cell=\\"\\" medium-8\\"=\\"\\"><h3>Beware of complex inputs</h3>
<p>Don't just think about code complexity, think about <em>input</em> complexity.↵↵Computers only understand numbers,
<a href=\\"/articles/beware-complex-inputs\\">Full Article</a></p></div><div class=\\"blurb-art\\" cell=\\"\\" medium-3\\"=\\"\\"><div class=\\"vhp-art-snowflake\\"></div></div><div class=\\"blurb-text\\" cell=\\"\\" medium-8\\"=\\"\\"><h3>We Love Open Questions</h3>
<p>Why do we talk about things that have not been \\"proven\\" in research?↵↵Because science is a work in progress.
<a href=\\"/articles/open-questions\\">Full Article</a></p></div><div class=\\"blurb-art\\" cell=\\"\\" medium-3\\"=\\"\\"><div class=\\"vhp-art-question\\"></div></div><div class=\\"blurb-text\\" cell=\\"\\" medium-8\\"=\\"\\"><h3>Always Be Refactoring</h3>
<p>You can't secure it if you can't maintain it.↵↵<strong>Refactoring</strong> is the practice of improving the maintainability
<a href=\\"/articles/refactoring\\">Full Article</a></p></div><div class=\\"blurb-art\\" cell=\\"\\" medium-3\\"=\\"\\"><div class=\\"vhp-art-refactor\\"></div></div><hr><hr><hr>"
`;

exports[`Display All Articles as One Size 1`] = `
"<div class=\\"blurb-text\\" cell=\\"\\" medium-8\\"=\\"\\"><h3>Beware of complex inputs</h3>
<p>Don't just think about code complexity, think about <em>input</em> complexity.↵↵Computers only understand numbers,
<a href=\\"/articles/beware-complex-inputs\\">Full Article</a></p></div><div class=\\"blurb-art\\" cell=\\"\\" medium-3\\"=\\"\\"><div class=\\"vhp-art-snowflake\\"></div></div><div class=\\"blurb-text\\" cell=\\"\\" medium-8\\"=\\"\\"><h3>We Love Open Questions</h3>
<p>Why do we talk about things that have not been \\"proven\\" in research?↵↵Because science is a work in progress.
<a href=\\"/articles/open-questions\\">Full Article</a></p></div><div class=\\"blurb-art\\" cell=\\"\\" medium-3\\"=\\"\\"><div class=\\"vhp-art-question\\"></div></div><div class=\\"blurb-text\\" cell=\\"\\" medium-8\\"=\\"\\"><h3>Always Be Refactoring</h3>
<p>You can't secure it if you can't maintain it.↵↵<strong>Refactoring</strong> is the practice of improving the maintainability
<a href=\\"/articles/refactoring\\">Full Article</a></p></div><div class=\\"blurb-art\\" cell=\\"\\" medium-3\\"=\\"\\"><div class=\\"vhp-art-refactor\\"></div></div><hr><hr><hr>"
`;
